libmachine/provision/photon.go

```diff
@@ -0,0 +1,194 @@
+package provision
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"github.com/docker/machine/drivers"
+	"github.com/docker/machine/libmachine/auth"
+	"github.com/docker/machine/libmachine/engine"
+	"github.com/docker/machine/libmachine/provision/pkgaction"
+	"github.com/docker/machine/libmachine/swarm"
+	"github.com/docker/machine/log"
+	"github.com/docker/machine/utils"
+)
+
+func init() {
+	Register("Photon", &RegisteredProvisioner{
+		New: NewPhotonProvisioner,
+	})
+}
+
+func NewPhotonProvisioner(d drivers.Driver) Provisioner {
+	return &PhotonProvisioner{
+		GenericProvisioner{
+			DockerOptionsDir:  "/etc/docker",
+			DaemonOptionsFile: "/etc/systemd/system/docker.service",
+			OsReleaseId:       "photon",
+			Driver:            d,
+		},
+	}
+}
+
+type PhotonProvisioner struct {
+	GenericProvisioner
+}
+
+func (provisioner *PhotonProvisioner) Service(name string, action pkgaction.ServiceAction) error {
+	reloadDaemon := false
+	switch action {
+	case pkgaction.Start, pkgaction.Restart:
+		reloadDaemon = true
+	}
+
+	// systemd needs reloaded when config changes on disk; we cannot
+	// be sure exactly when it changes from the provisioner so
+	// we call a reload on every restart to be safe
+	if reloadDaemon {
+		if _, err := provisioner.SSHCommand("sudo systemctl daemon-reload"); err != nil {
+			return err
+		}
+	}
+
+	command := fmt.Sprintf("sudo systemctl %s %s", action.String(), name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *PhotonProvisioner) Package(name string, action pkgaction.PackageAction) error {
+	var (
+		packageAction  string
+		updateMetadata = true
+	)
+
+	switch action {
+	case pkgaction.Install:
+		packageAction = "install"
+	case pkgaction.Remove:
+		packageAction = "erase"
+		updateMetadata = false
+	case pkgaction.Upgrade:
+		packageAction = "upgrade"
+	}
+
+	if updateMetadata {
+		if _, err := provisioner.SSHCommand("sudo -E tdnf --assumeyes makecache"); err != nil {
+			return err
+		}
+	}
+
+	command := fmt.Sprintf("sudo -E tdnf --assumeyes %s %s", packageAction, name)
```

Comment:
Right now AppCatalyst TP1 (technical preview 1) ships with Docker 1.7.0 preinstalled, `tdnf` is a replacement for `yum` (which is being replaced by `dnf`, `tdnf` is tiny-`dnf`).

`tdnf` pulls new packages from the photon repository who is curated by the photon team at VMware.